src/sqlfluff/rules/L003.py

```diff
@@ -1,15 +1,180 @@
-from typing import List, Optional, Sequence, Tuple
+import dataclasses
+import itertools
+from typing import Dict, List, Optional, Sequence, Tuple
 
 from sqlfluff.core.parser import WhitespaceSegment
-from sqlfluff.core.parser.segments import BaseSegment, RawSegment
+from sqlfluff.core.parser.segments import BaseSegment
 from sqlfluff.core.rules.functional import rsp, sp, Segments
 from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
 from sqlfluff.core.rules.doc_decorators import (
     document_fix_compatible,
     document_configuration,
 )
 from sqlfluff.core.templaters import TemplatedFile
+from sqlfluff.core.templaters.base import RawFileSlice
+
+
+@dataclasses.dataclass
+class _LineSummary:
+    """A dataobject to represent a line.
+
+    Doubles as a memory object (holding state)
+    that can be converted to a fresh summary instance
+    """
+
+    line_no: int = 0
+    templated_line: Optional[int] = None
+    line_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    indent_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    indent_size: int = 1
+
+    # The final balance of a line once fully considered
+    indent_balance: int = 0
+    # The indent as it was once we say this lines "Anchor"
+    anchor_indent_balance: int = 0
+    hanging_indent: Optional[int] = None
+    clean_indent: bool = True
+    templated_line_type: Optional[str] = None
+    is_comment_line: bool = False
+    is_empty_line: bool = False
+    line_anchor: Optional[BaseSegment] = None
+    # These are only required for carrying state between lines
+    line_indent_stack: List[int] = dataclasses.field(default_factory=list)
+    hanger_pos: Optional[int] = None
+
+    def __repr__(self) -> str:
+        """Printable Summary without Segments."""
+        keys_to_strip = (
+            "line_buffer",
+            "indent_buffer",
+            "as_of_anchor",
+        )
+        print_dict: Dict = {
+            key: value
+            for key, value in self.__dict__.copy().items()
+            if key not in keys_to_strip
+        }
+
+        return print_dict.__repr__()
+
+    @property
+    def template_content(self):  # pragma: no cover
+        return "".join(
+            seg.raw or getattr(seg, "source_str", "") for seg in self.line_buffer
+        )
+
+    def memory_line_reset(self):
+        """When acting like a MemoryLine this resets state."""
+        self.templated_line = False
+        self.indent_buffer = []
+        # N.B the clean indent depends on the buffer of the line just pased
+        self.clean_indent = _is_clean_indent(self.line_buffer)
+        self.line_buffer = []
+        self.line_indent_stack = []
+        self.indent_size = 0
+        self.hanger_pos = None
+        self.line_anchor = None
+        self.templated_line_type = None
+        return self
+
+    def from_memo_line(self, line_no: int, templated_file: Optional[TemplatedFile]):
```

Comment:
Does "memo" here mean the same thing as "memory line" in the previous function?
```suggestion
    def from_memory_line(self, line_no: int, templated_file: Optional[TemplatedFile]):
```